src/Tools/AnalyzerRunner/CodeRefactoringRunner.cs

```diff
@@ -0,0 +1,256 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Composition;
+using static AnalyzerRunner.Program;
+
+namespace AnalyzerRunner
+{
+    internal sealed class CodeRefactoringRunner
+    {
+        private readonly Options _options;
+        private readonly ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>> _refactorings;
+        private readonly ImmutableDictionary<string, ImmutableHashSet<int>> _syntaxKinds;
+
+        public CodeRefactoringRunner(Options options)
+        {
+            _options = options;
+
+            var refactorings = GetCodeRefactoringProviders(options.AnalyzerPath);
+            _refactorings = FilterRefactorings(refactorings, options);
+
+            _syntaxKinds = GetSyntaxKinds(options.RefactoringNodes);
+        }
+
+        public bool HasRefactorings => _refactorings.Any(pair => pair.Value.Any());
+
+        public async Task RunAsync(Workspace workspace, CancellationToken cancellationToken)
+        {
+            if (!HasRefactorings)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(_options.ProfileRoot))
+            {
+                ProfileOptimization.StartProfile(nameof(CodeRefactoringRunner));
+            }
+
+            var solution = workspace.CurrentSolution;
+            var stopwatch = Stopwatch.StartNew();
+
+            var updatedSolution = solution;
+
+            foreach (var project in solution.Projects)
+            {
+                foreach (var document in project.Documents)
+                {
+                    var newDocument = await RefactorDocumentAsync(document, cancellationToken).ConfigureAwait(false);
+                    if (newDocument is null)
+                    {
+                        continue;
+                    }
+
+                    updatedSolution = updatedSolution.WithDocumentSyntaxRoot(document.Id, await newDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false));
+                }
+            }
+
+            if (_options.ApplyChanges)
+            {
+                workspace.TryApplyChanges(updatedSolution);
+            }
+        }
+
+        private async Task<Document> RefactorDocumentAsync(Document document, CancellationToken cancellationToken)
+        {
+            var syntaxKinds = _syntaxKinds[document.Project.Language];
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var node in root.DescendantNodesAndTokens(descendIntoTrivia: true))
+            {
+                if (!syntaxKinds.Contains(node.RawKind))
```

Comment:
Why do we need the syntax kind check given you already have refactorings grouped by language?

Reply:
It only attempts to run the refactoring on nodes and tokens that are requested. For example, the arguments I used for the recursive patterns refactoring are:

> $(OutDir) $(SolutionDir)Roslyn.sln /stats /a CSharpUseRecursivePatternsCodeRefactoringProvider /refactor EqualsEqualsToken /refactor AmpersandAmpersandToken /refactor WhenKeyword /apply